src/main/java/gndata/lib/util/ChangeHelper.java

```diff
@@ -12,46 +13,195 @@
 
-        if (infm.validate().isValid()) {
-            tracker.logChange(ch);
-        } else {
-            ch.undoFrom(model);
+        if (!(position > -1)) {
+            throw new IllegalStateException("No previous changes recorded");
         }
+
+        if (isLastUpdate()) {  // this is an update
+            Change toRemove = changes.get(position);
+            Change toAdd = changes.get(position - 1);
+
+            model.remove(toRemove.getChange());
+            position -= 1;
+
+            model.add(toAdd.getChange());
+            position -= 1;
+
+        } else {  // these are non-update operations
+            Change ch = changes.get(position);
+
+            if (ch.isPositive()) {
+                model.remove(ch.getChange());
+            } else {
+                model.add(ch.getChange());
+            }
+
+            position -= 1;
+        }
+
+        if (validate) {
+            if (!validate().isValid()) {
+                redo();
+                throw new IllegalStateException("Undo leads to inconsistent state");
+            }
+        }
+
     }
 
-    /*   Public interface   */
+    public void redo() {
+        if (!(position < changes.size() - 1)) {
+            throw new IllegalStateException("Nothing to re-do");
+        }
+
+        if (isNextUpdate()) {  // this is an update
+            Change toRemove = changes.get(position + 1);
+            Change toAdd = changes.get(position + 2);
+
+            model.remove(toRemove.getChange());
+            position += 1;
+
+            model.add(toAdd.getChange());
+            position += 1;
+
+        } else {  // these are non-update operations
+            Change ch = changes.get(position + 1);
+
+            if (ch.isPositive()) {
+                model.add(ch.getChange());
+            } else {
+                model.remove(ch.getChange());
+            }
+
+            position += 1;
+        }
 
-    public void create(Model NewObject) {
-        Change ch = new Insert(NewObject);
-        applyChange(ch);
+        if (validate) {
+            if (!validate().isValid()) {
+                undo();
+                throw new IllegalStateException("Redo leads to inconsistent state");
+            }
+        }
     }
 
-    public void update(Model newObject) {
-        Change ch = new Insert(newObject);
-        applyChange(ch);
+    public Change get(int index) {
+        return changes.get(index);
     }
 
-    public void delete(String uri) {
-        Change ch = new Delete(uri);
-        applyChange(ch);
+    public int size() {
+        return changes.size();
     }
 
-    // TODO add more methods
+    /*   Private helper functions   */
+
+    private static boolean isUpdate(Change removed, Change added) {
+
+        // assume update is always first remove and then add
+
+        if (added.getChange().size() != removed.getChange().size()) {
+            return false;
+        }
+
+        StmtIterator iter = removed.getChange().listStatements();
```

Comment:
Please add a comment to the code, that we know about the special case when a user has manually deleted and added property, but that we deliberately chose to treat it as a single update event for now.